feat(mojaloop/#3400): benchmarking performance for ALS

mdebarros opened the pull request:
feat(mojaloop/#3400): benchmarking performance for ALS - https://github.com/mojaloop/project/issues/3400
- created documentation, and scenarios for characterizing ALS
- documented performance data from ALS characterization runs

feat(mojaloop/#3424): analyse als perf results - https://github.com/mojaloop/project/issues/3424
- Added analysis for scenarios 1-16

mdebarros committed:
added initial template, and scneario runs for 20230726

mdebarros committed:
updates

mdebarros committed:
added exports for test-scenarios

mdebarros committed:
added environment

mdebarros committed:
added log files

mdebarros committed:
added test scenario s5-1690447500991

mdebarros committed:
added data to s5-1690447500991

mdebarros committed:
added 20230727/s7-1690407403663

mdebarros committed:
moved s7-1690407403663/ to 20230726 folder

vijayg10 committed:
feat: added two scenarios

mdebarros committed:
updates to main readme, and template

vijayg10 committed:
feat: added s9 scenario results

vijayg10 committed:
Merge branch 'feat/performance-maintenance-characterisation-als' of https://github.com/mojaloop/ml-perf-characterization into feat/performance-maintenance-characterisation-als

mdebarros committed:
added more test-cases

kleyow committed:
chore: add test results for #11-13

kleyow committed:
chore: remove

mdebarros committed:
fleshed out main readme with approach, and added drawio diagrams

mdebarros committed:
fleshed out main readme with approach, and added drawio diagrams

mdebarros committed:
fleshed out main readme with approach, and added drawio diagrams

mdebarros committed:
fleshed out main readme with approach, and added drawio diagrams

vijayg10 committed:
feat: captured s14 data

vijayg10 committed:
Merge branch 'feat/performance-maintenance-characterisation-als' of https://github.com/mojaloop/ml-perf-characterization into feat/performance-maintenance-characterisation-als

mdebarros committed:
added docker version

mdebarros committed:
Merge branch 'feat/performance-maintenance-characterisation-als' of github.com:mojaloop/ml-perf-characterization into feat/performance-maintenance-characterisation-als

vijayg10 committed:
feat: added empty folder structures for s15 and s16 scenarios

kleyow committed:
chore: add images for s15 and s16

sri-miriyala commented on 20230726/s2-1690376653994/README.md:
```diff
@@ -0,0 +1,57 @@
+# Scenario 2 - ALS Baseline with Sims-only
```
I think it will be more meaningful if we change it to "ALS Baseline with Sims". InScenario 1 as it was only Sims without ALS and in this scenario the baseline is with ALS.

mdebarros replied:
Noted.

sri-miriyala commented on 20230726/s6-1690380087112/README.md:
```diff
@@ -0,0 +1,61 @@
+# Scenario 6 - ALS Baseline ALS Baseline with Sims-only, HTTP-Keep-Alive enabled
```
Repeat of ALS Baseline and request to update from Sims-only to Sims.

mdebarros replied:
Noted

sri-miriyala commented on 20230726/s7-1690407403663/README.md:
```diff
@@ -0,0 +1,60 @@
+# Scenario 1 - ALS-bypass Baseline with Sims-only
```
Is this the right description for this scenario? This is for Scenario 1

mdebarros replied:
Nice catch.

This should be `ALS Baseline with Sims-only, UV_THREADS`, I have fixed this and also renamed it to match you previous suggestions, i.e. `Sims-only` -> `Sims` = `ALS Baseline with Sims, UV_THREADS`.

sri-miriyala commented on 20230727/s10-1690466917636/README.md:
```diff
@@ -0,0 +1,64 @@
+# Scenario 8 - ALS Baseline with Sims-only, multiple k6 VUs
```
This README is in s10 folder. Based on the pattern before, each folder name was representing corresponding scenario number. Here Scenario 8 is in s-10 folder.

mdebarros replied:
Fixed

sri-miriyala commented on 20230727/s12-1690491686694/README.md:
```diff
@@ -0,0 +1,66 @@
+# Scenario 12 - ALS Baseline with Sims-only, Disabled JSON.stringify. ALS v14.2.3 + ALS Scale 2.
```
Can we have a consistent naming convention for all the scenarios? With high level scenario description, ALS version, ALS scale, K6 VU, Comments across all the scenarios?

sri-miriyala commented on 20230727/s14-1690504862678/README.md:
```diff
@@ -0,0 +1,68 @@
+# Scenario 14 - ALS Baseline with Sims-only, Disabled JSON.stringify ALS v14.2.3 + Scale 4 + 6x k6 VUs
```
Can we update this to: ALS Scale - 4, K6 VUs - 6

sri-miriyala commented on 20230727/s11-1690488145504/README.md:
```diff
@@ -0,0 +1,66 @@
+# Scenario 11 - ALS Baseline with Sims-only, Disabled JSON.stringify. ALS v14.2.3 + 4x k6 VUs.
```
Can we add ALS scale.

mdebarros replied:
The ALS was not scaled here, only the K6 VUs were increased to 4 from 1. Does that makes sense?

sri-miriyala commented on 20230727/s13-1690493569083/README.md:
```diff
@@ -0,0 +1,65 @@
+# Scenario 13 - ALS Baseline with Sims-only, Disabled JSON.stringify. ALS v14.2.3 + + Scale 4 + 4x k6 VUs.
```
Can we change it to ALS scale - 4

sri-miriyala commented on 20230727/s5-1690447500991/README.md:
```diff
@@ -0,0 +1,59 @@
+# Scenario 1 - ALS-bypass Baseline with Sims-only
```
Is this the right description for this scenario?

mdebarros replied:
Fixed.

sri-miriyala commented on 20230727/s8-1690457241591/README.md:
```diff
@@ -0,0 +1,62 @@
+# Scenario 8 - ALS Baseline with Sims-only, multiple k6 VUs
```
The K6 VU count is dynamic, not static?

mdebarros replied:
Its static.

We change in between runs. We actually ran this scenario increasing the VUs from 1 to 5.

mdebarros replied:
There was no substantial difference between 1-5, thus why we only captured one of the results

mdebarros replied:
Also fixed title.

sri-miriyala commented on 20230727/s9-1690470790793/README.md:
```diff
@@ -0,0 +1,66 @@
+# Scenario 8 - ALS Baseline with Sims-only, multiple k6 VUs
```
Is this the right description for Scenario 9?

mdebarros replied:
Fixed.

sri-miriyala commented on 20230728/s15-1690548412616/README.md:
```diff
@@ -0,0 +1,69 @@
+# Scenario 15 - ALS Baseline with Sims-only, Disabled JSON.stringify [ALS v14.2.3](https://github.com/mojaloop/account-lookup-service/releases/tag/v14.2.3) + Scale 4 + 6x k6 VUs + Enhanced Caching for ValidateParticipants | 1 | N | [account-lookup-service/pull/461](https://github.com/mojaloop/account-lookup-service/pull/461), [v14.2.4-snapshot.3](https://github.com/mojaloop/account-lookup-service ...
```
Can this description be consistent with the rest of the scenario descriptions? What does  | 1 | N | represent?

sri-miriyala commented on 20230728/s16-1690552431770/README.md:
```diff
@@ -0,0 +1,72 @@
+# Scenario 16 - ALS Baseline with Sims-only, Disabled JSON.stringify [ALS v14.2.3](https://github.com/mojaloop/account-lookup-service/releases/tag/v14.2.3) + Scale 4 + 6x k6 VUs + Enhanced Caching for ValidateParticipants + OracleRequest | 1 | N | [account-lookup-service/pull/461](https://github.com/mojaloop/account-lookup-service/pull/461), [v14.3.0-snapshot.0](https://github.com/mojaloop/accoun ...
```
What is different in this scenario from scenario 15 apart from different ALS version? It has Oracle Request. Does this mean there is Oracle lookup involved?

mdebarros replied:
Correct. 

The Oracle Request is being cached.

sri-miriyala reviewed (commented):
Added comments around consistency of scenario descriptions.

sri-miriyala commented on README.md:
```diff
@@ -0,0 +1,144 @@
+
+## Test Scenarios
+
+Scenario | Description | Test-Case | Repeatable (Y/N) | Notes
```
Can this table be updated with the changes in the scenario descriptions in the respective folders

sri-miriyala reviewed (commented):
Added a comment in the main README for the overall scenarios table.

mdebarros committed:
feat(mojaloop/#3424): analyse als perf results

feat(mojaloop/#3424): analyse als perf results - https://github.com/mojaloop/project/issues/3424
- Added analysis for scenarios 1-14

feat(mojaloop/#3400): benchmarking performance for als - https://github.com/mojaloop/project/issues/3400
- Updated comments based on PR review.

mdebarros committed:
Merge branch 'feat/performance-maintenance-characterisation-als' of github.com:mojaloop/ml-perf-characterization into feat/performance-maintenance-characterisation-als

mdebarros committed:
feat(mojaloop/#3424): analyse als perf results

feat(mojaloop/#3424): analyse als perf results - https://github.com/mojaloop/project/issues/3424
- Added analysis for scenarios 15-16

mdebarros committed:
fixes to headers for consistency

mdebarros committed:
fixed git.ignore --> .gitignore

vijayg10 commented on 20230727/s10-1690466917636/README.md:
```diff
@@ -0,0 +1,87 @@
+    - `Account-Lookup-Service` is showing a reduced CPU usage of `70%`, `30%` down from `100%`, indicating that the service is able to do more if K6 VUs are increased.
+    - `Callback-Handler` is within bounds of the `Scenario #1`.
+
+- `JSON.stringify` is a demanding operation which is causing a bottleneck on the NodeJS `Event-loop` impacting the End-to-end `Op/s` and `duration`.
```
It is also impacting HTTP Sync response which is a major issue than end-to-end duration. This eventually causing K8S pod restarts because of the http health endpoint unavailability because of event loop.

mdebarros replied:
Done

mdebarros committed:
fix: standardised test scenarios names

mdebarros committed:
fixed comment

mdebarros committed:
fixes to scratched out TSs

sri-miriyala commented on 20230727/s10-1690466917636/README.md:
```diff
@@ -0,0 +1,87 @@
+# Scenario 10 - FSPIOP Discovery GET Parties with ALS + Sims + JSON.Stringify fix - ALS:v14.2.3, scale:1, k6vu:1 ALS
```
Is ALS needed at the end @mdebarros

mdebarros replied:
Fixed

sri-miriyala commented on 20230727/s8-1690457241591/README.md:
```diff
@@ -0,0 +1,66 @@
+# Scenario 8 - FSPIOP Discovery GET Parties with ALS + Sims + K6 VU scaling - ALS:v14.2.2, scale:1, k6vu:1-5
```
@mdebarros ..k6vu 1-5 does it indicate that started with 1vu and increased till 5vus?

mdebarros replied:
Correct.

mdebarros committed:
fixed bad search-replace

mdebarros committed:
added story links

sri-miriyala reviewed (approved):
LGTM

mdebarros committed:
added follow-up story for end-to-end metric instrumentation on the ALS

mdebarros committed:
Updated documentation based on review comments.

mdebarros committed:
added k8s test case and scenario name/config
